message/messaging_service.cc

```diff
@@ -753,22 +776,21 @@ shared_ptr<messaging_service::rpc_protocol_client_wrapper> messaging_service::ge
             return false;
         }
 
-        if (_cfg.compress == compress_what::all || !has_topology()) {
+        // See comment above `TOPOLOGY_INDEPENDENT_IDX`.
+        if (_cfg.compress == compress_what::all || idx == TOPOLOGY_INDEPENDENT_IDX) {
             return true;
         }
 
-        return !is_same_dc(id.addr);
+        return !has_topology() || !is_same_dc(id.addr);
     }();
 
     auto must_tcp_nodelay = [&] {
-        if (idx == 0) {
```

Comment:
This check has nothing to do with "topology independenc~~e da~~y ". It's just to make gossiper (see e.g. #11490) work smoother

Reply:
I mention tcp_nodelay at the TOPOLOGY_INDEPENDENT_IDX comment too.